File: src/main/other/advent_of_code_2022/src/day4/part2.rs
extern crate test;

use std::fs::File;
use std::io::{BufRead, BufReader};

use itertools::Itertools;

use crate::day4::interval::Interval;

pub fn part2() -> Result<u32, Box<dyn std::error::Error>> {
    let file = File::open("input/day4/input.txt")?;
    let reader = BufReader::new(file);

    let mut res = 0;
    for line in reader.lines() {
        let line = line.unwrap();
        let (first, second) = line
            .split(',')
            .map(|str| Interval::from_string(str))
            .next_tuple()
            .unwrap();

        if first.contains_part(&second) {
            res += 1
        }
    }

    Ok(res)
}

#[cfg(test)]
mod tests {
    use super::*;
    use test::Bencher;
    #[test]
    fn test_part2() {
        assert_eq!(835, part2().unwrap());
    }

    #[bench]
    fn bench_part2(b: &mut Bencher) {
        b.iter(|| {
            part2().unwrap();
        })
    }
}

File: src/main/other/advent_of_code_2023/day4/part2.rb
input = File.open(File.join(File.dirname(__FILE__), "input.txt"), "r").map(&:chomp)

copies = Array.new(input.size, 1)
res = input.each_with_index.sum do |line, line_index|
    line.split('|').map { |s| s.scan(/\d+/).map(&:to_i) }.then.sum do |first, second|
        (first[1...] & second).size.times do |dy|
            copies[line_index + dy + 1] += copies[line_index]
        end
        copies[line_index]
    end
end

pp res
